### PR TITLE
feat(virtualnetwork): firewall manager security configuration 

### LINKED ISSUE #137: feat: add internet security for vhub connections
- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

## Description

https://learn.microsoft.com/en-us/azure/templates/microsoft.network/virtualhubs/hubvirtualnetworkconnections?pivots=deployment-language-bicep

Add `enableInternetSecurity` bool

[AB#25716](https://dev.azure.com/CSUSolEng/2a9d5a5a-8998-4ad0-81c8-ef3045e4da97/_workitems/edit/25716)

**Is your feature request related to an issue?**

### PR DESCRIPTION
**Pull request by b0bu**
# Overview/summary

Replace this with a brief description of what this Pull Request fixes, changes, etc.

## This PR fixes/adds/changes/removes

fixes #137 

1. Add toggles to configure a managed vnet peers security configuration. Due to the different states that can be toggled I've just added basic "and/and not" logic to handle that i.e. `secure_internet_traffic = true` results in a `public_traffic` route `secure_private_traffic = true` results in a `private_traffic` route and both together result in an `all_traffic` route which merges both changes into one. 

```hcl
      vwan_security_configuration = {
        secure_internet_traffic = true
        secure_private_traffic  = true
        next_hop = "/subscriptions/0000/resourceGroups/Connectivity/providers/Microsoft.Network/azureFirewalls/fw-uksouth"
      }
```
### Breaking changes

1. None so far

## Testing evidence

Please provide testing evidence to show that your Pull Request works/fixes as described and documented above.

I haven't written any tests yet but I have built out the infrastructure with the various toggle combinations ensuring it behaves as expected.

## As part of this pull request I have

- [x] Checked for duplicate [Pull Requests](https://github.com/Azure/terraform-azurerm-lz-vending/pulls)
- [x] Associated it with relevant [issues](https://github.com/Azure/terraform-azurerm-lz-vending/issues), for tracking and closure.
- [x] Run and `make fmt` & `make docs` to format your code and update documentation.
- [x] Created unit and deployment tests and provided evidence.
- [ ] Updated relevant and associated documentation.

### CODE REVIEW

**Comment by matt-FFFFFF** [comment]
Thank you for the PR @b0bu!

I'll get around to reviewing in the coming days. But initial feedback:

* The variables descriptions need updating
* The docs will need generating (make docs && make fmt from a WSL/Linux shell)
* Tests need creating

I'll help with any of it but if you want to take on anything then please do 😄

**Comment by b0bu** [comment]
Hello there @matt-FFFFFF yea was planning on it! I'll take care of some of these today. And I agree about the naming. Any suggestions would be helpful.

**Comment by matt-FFFFFF** [comment]
Hey @b0bu

Before you go much further, after reviewing in more detail, I have two things. I'll start a formal review so it's clear what I'm referring to.

File: modules/virtualnetwork/main.tf [hunk 1]
```diff
@@ -155,10 +156,96 @@ resource "azapi_resource" "vhubconnection" {
...
+          labels = each.value.vwan_security_configuration.secure_private_traffic == true ? ["none"] : local.vwan_propagated_routetables_labels[each.key]
         }
       }
     }
   })
 }
+
+# azapi_update_resource.vhubdefaultroutetableinternettraffic creates public_traffic route which associated vnets then learn
+# with a next hop route for the internet as the vhub's azure firewall
+resource "azapi_update_resource" "vhubdefaultroutetableinternettraffic" {
```
Comment (matt-FFFFFF) [inline modules/virtualnetwork/main.tf:169]: I don't think we should update hub route tables using this module.

1. This does not scale, with 100s of instances of this module the number of API calls will cause rate limiting.
2. I do not believe this module is the right place to do this. Any hub route table config should be performed using the terraform code that deploys the vwan hub resource.

File: modules/virtualnetwork/main.tf [hunk 2]
```diff
@@ -155,10 +156,96 @@ resource "azapi_resource" "vhubconnection" {
...
+          "nextHop"     = each.value.vwan_security_configuration.next_hop
+        }
+      ]
+    }
+  })
+}
+
+# azapi_update_resource.vhubdefaultroutetableprivatetraffic creates private_traffic route which associated vnets then learn
+# with a next hop route for martian prefix list as the vhub's azure firewall
+resource "azapi_update_resource" "vhubdefaultroutetableprivatetraffic" {
```
Comment (matt-FFFFFF) [inline modules/virtualnetwork/main.tf:196]: See comment on L169

File: modules/virtualnetwork/main.tf [hunk 3]
```diff
@@ -155,10 +156,96 @@ resource "azapi_resource" "vhubconnection" {
...
+          "nextHop"     = each.value.vwan_security_configuration.next_hop
+        }
+      ]
+    }
+  })
+}
+
+# azapi_update_resource.vhubdefaultroutetablealltraffic creates private_traffic route which associated vnets then learn
+# with a next hop route for the internet and martian prefix list as the vhub's azure firewall
+resource "azapi_update_resource" "vhubdefaultroutetablealltraffic" {
```
Comment (matt-FFFFFF) [inline modules/virtualnetwork/main.tf:225]: See comment on L169

File: modules/virtualnetwork/main.tf [hunk 4]
```diff
@@ -155,10 +156,96 @@ resource "azapi_resource" "vhubconnection" {
           id = each.value.vwan_associated_routetable_resource_id != "" ? each.value.vwan_associated_routetable_resource_id : "${each.value.vwan_hub_resource_id}/hubRouteTables/defaultRouteTable"
         }
         propagatedRouteTables = {
-          ids    = local.vwan_propagated_routetables_resource_ids[each.key]
-          labels = local.vwan_propagated_routetables_labels[each.key]
+          ids    = each.value.vwan_security_configuration.secure_private_traffic == true ? local.vwan_propagated_noneroutetables_resource_ids[each.key] : local.vwan_propagated_routetables_resource_ids[each.key]
```
Comment (matt-FFFFFF) [inline modules/virtualnetwork/main.tf:159]: Please can you explain what this does? You are specifying the `noneRouteTable` in propagatedRouteTables when `secure_private_traffic` is enabled. The default behaviour was to use `defaultRouteTable`.

Why is this necessary when using secure_private_traffic?

Thanks

**Review by matt-FFFFFF — CHANGES_REQUESTED** [review]

File: modules/virtualnetwork/main.tf (hunk 1 above)
Comment (b0bu) [reply modules/virtualnetwork/main.tf:169]: I had similar thoughts while testing.

File: modules/virtualnetwork/main.tf (hunk 4 above)
Comment (b0bu) [reply modules/virtualnetwork/main.tf:159]: In security configuration when private traffic is set to "Bypass Azure Firewall"
Screenshot 2023-01-26 at 10 06 00
The connection object looks like this 
```json
{
  "name": "vhc-0e2426dc-d34c-5bb5-8b01-8dc1a999084d",
  "id": "/subscriptions/64db9221-622e-4f35-8c66-0000/resourceGroups/Connectivity/providers/Microsoft.Network/virtualHubs/vhub-uksouth/hubVirtualNetworkConnections/vhc-0e2426dc-d34c-5bb5-8b01-8dc1a999084d",
  "etag": "W/\"bbde0027-4cc5-4599-898f-3b6e415fb4ad\"",
  "type": "Microsoft.Network/virtualHubs/hubVirtualNetworkConnections",
  "properties": {
    "provisioningState": "Succeeded",
    "resourceGuid": "843c14ab-a070-457a-a085-0000",
    "routingConfiguration": {
      "associatedRouteTable": {
        "id": "/subscriptions/64db9221-622e-4f35-8c66-0000/resourceGroups/Connectivity/providers/Microsoft.Network/virtualHubs/vhub-uksouth/hubRouteTables/defaultRouteTable"
      },
      "propagatedRouteTables": {
        "labels": [
          "default"
        ],
        "ids": [
          {
            "id": "/subscriptions/64db9221-622e-4f35-8c66-0000/resourceGroups/Connectivity/providers/Microsoft.Network/virtualHubs/vhub-uksouth/hubRouteTables/defaultRouteTable"
          }
        ]
      },
      "vnetRoutes": {
        "staticRoutes": []
      }
    },
    "remoteVirtualNetwork": {
      "id": "/subscriptions/bb947477-068c-42c6-ac2c-0000/resourceGroups/connectivity-uks/providers/Microsoft.Network/virtualNetworks/vnet_testing_uks"
    },
    "allowHubToRemoteVnetTransit": true,
    "allowRemoteVnetToUseHubVnetGateways": true,
    "enableInternetSecurity": false
  }
}
```
Where `propagatedRouteTables` labels and ids are `default` and `defaultRouteTable` respectively.

When secure private traffic is enabled for the api updates the connection object, it changes.
Screenshot 2023-01-26 at 10 21 33
and now looks like this
```json
{
  "name": "vhc-0e2426dc-d34c-5bb5-8b01-8dc1a999084d",
  "id": "/subscriptions/64db9221-622e-4f35-8c66-0000/resourceGroups/Connectivity/providers/Microsoft.Network/virtualHubs/vhub-uksouth/hubVirtualNetworkConnections/vhc-0e2426dc-d34c-5bb5-8b01-8dc1a999084d",
  "etag": "W/\"af2e1f66-25a1-439e-9873-56daf33fdcd9\"",
  "type": "Microsoft.Network/virtualHubs/hubVirtualNetworkConnections",
  "properties": {
    "provisioningState": "Updating",
    "resourceGuid": "843c14ab-a070-457a-a085-0000",
    "routingConfiguration": {
      "associatedRouteTable": {
        "id": "/subscriptions/64db9221-622e-4f35-8c66-0000/resourceGroups/Connectivity/providers/Microsoft.Network/virtualHubs/vhub-uksouth/hubRouteTables/defaultRouteTable"
      },
      "propagatedRouteTables": {
        "labels": [
          "none"
        ],
        "ids": [
          {
            "id": "/subscriptions/64db9221-622e-4f35-8c66-0000/resourceGroups/Connectivity/providers/Microsoft.Network/virtualHubs/vhub-uksouth/hubRouteTables/noneRouteTable"
          }
        ]
      },
      "vnetRoutes": {
        "staticRoutes": []
      }
    },
    "remoteVirtualNetwork": {
      "id": "/subscriptions/bb947477-068c-42c6-ac2c-0000/resourceGroups/connectivity-uks/providers/Microsoft.Network/virtualNetworks/vnet_testing_uks"
    },
    "allowHubToRemoteVnetTransit": true,
    "allowRemoteVnetToUseHubVnetGateways": true,
    "enableInternetSecurity": false
  }
```
Where `propagatedRouteTables` labels and ids become `none` and `noneRouteTable` respectively. The code is literally just saying 

```bash
if secure_traffic is true
  labels and ids = none and noneRouteTable
else 
  labels and ids = default and defaultRouteTable
```
To answer more specifically your question as to *why* I'm not sure, I'd speculate that since traffic now has to route via the fw that how the routes are advertised has to change.

Comment (b0bu) [reply modules/virtualnetwork/main.tf:159]: So propagating to none from what I can see essentially isolates the vnet where its routes aren't published to the hub. I'd asusme this is how the dynamic managed routing of vwan is turned off, where it's only reachable via the fw router and fw policy.

Comment (matt-FFFFFF) [reply modules/virtualnetwork/main.tf:159]: Thanks @b0bu, this makes sense. I will double check but think I'm happy with this :) Thanks for your efforts here

File: modules/virtualnetwork/main.tf (hunk 1 above)
Comment (matt-FFFFFF) [reply modules/virtualnetwork/main.tf:169]: Thanks - I'm happy to remove this but if you want to then that would be helpful.

File: modules/virtualnetwork/main.tf (hunk 4 above)
Comment (b0bu) [reply modules/virtualnetwork/main.tf:159]: As an update for anyone that finds this PR, I added the routing config to my secure-vhub module with options which allows me to set the routing globally at the firewall. Now all I need is to get this merged so we can enable it at the connection!  

```hcl
module "secure-vhub-ukwest" {
  ...
  secure_private_traffic = false
  secure_public_traffic  = false
  ...
}
```
where module contents has
```hcl
variable "secure_private_traffic" {
   type = bool
   default = true
}
variable "secure_public_traffic" {
   type = bool
   default = true
}

resource azurerm_virtual_hub_route_table_route "private" {
  for_each = var.secure_private_traffic && !var.secure_public_traffic ? toset(["1"]) : []
  route_table_id = azurerm_virtual_hub.vhub.default_route_table_id

  name             = "private_traffic"
  destination_type = "CIDR"
  destinations     = ["10.0.0.0/8","172.16.0.0/12","192.168.0.0/16"]
  next_hope_type   = "ResourceId"
  next_hop         = azure_firewall.fw.id 
}

resource azurerm_virtual_hub_route_table_route "public" {
  for_each = var.secure_public_traffic && !var.secure_private_traffic ? toset(["1"]) : []
  route_table_id = azurerm_virtual_hub.vhub.default_route_table_id

  name             = "public_traffic"
  destination_type = "CIDR"
  destinations     = ["0.0.0.0/0"]
  next_hope_type   = "ResourceId"
  next_hop         = azure_firewall.fw.id 
}

resource azurerm_virtual_hub_route_table_route "all" {
  for_each = var.secure_private_traffic && var.secure_public_traffic ? toset(["1"]) : []
  route_table_id = azurerm_virtual_hub.vhub.default_route_table_id

  name             = "all_traffic"
  destination_type = "CIDR"
  destinations     = ["0.0.0.0/0", "10.0.0.0/8","172.16.0.0/12","192.168.0.0/16"]
  next_hope_type   = "ResourceId"
  next_hop         = azure_firewall.fw.id 
}
```

**Comment by matt-FFFFFF** [comment]
Thanks for the updates @b0bu 

I still need to update the tests but this is looking good.

**Comment by matt-FFFFFF** [comment]
@b0bu please can you do this so that I can eventually merge your work?

You type in a comment @ mentioning microsoft-github-policy-service and type agree

> @microsoft-github-policy-service agree [company="{your company}"]

Thanks!

> @b0bu please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> 
> Contributor License Agreement

**Comment by b0bu** [comment]
@matt-FFFFFF 

Let me push some tests and you can review :)

@microsoft-github-policy-service agree

**Comment by matt-FFFFFF** [comment]
> @matt-FFFFFF
> 
> Let me push some tests and you can review :)
> 
> @microsoft-github-policy-service agree

Thanks!

I am refactoring the tests at the moment, they are a little cumbersome right now so apologies.

File: modules/virtualnetwork/variables.tf [hunk 5]
```diff
@@ -65,6 +65,11 @@ variable "virtual_networks" {
     vwan_hub_resource_id                     = optional(string, "")
     vwan_propagated_routetables_labels       = optional(list(string), [])
     vwan_propagated_routetables_resource_ids = optional(list(string), [])
+    vwan_security_configuration = object({
```
Comment (matt-FFFFFF) [inline modules/virtualnetwork/variables.tf:68]: please can you wrap the vwan_security_configuration in an `optional()` with an empty object as default?

File: variables.virtualnetwork.tf [hunk 6]
```diff
@@ -51,6 +51,11 @@ variable "virtual_networks" {
     vwan_hub_resource_id                     = optional(string, "")
     vwan_propagated_routetables_labels       = optional(list(string), [])
     vwan_propagated_routetables_resource_ids = optional(list(string), [])
+    vwan_security_configuration = object({
```
Comment (matt-FFFFFF) [inline variables.virtualnetwork.tf:54]: same here, please wrap in `optional()` with empty object as default.

**Review by matt-FFFFFF — CHANGES_REQUESTED** [review]

**Comment by b0bu** [comment]
@matt-FFFFFF 
> I am refactoring the tests at the moment, they are a little cumbersome right now so apologies.

All good! it's been a busy week just getting back to this

**Comment by b0bu** [comment]
More updates coming to account for the new variable inputs, had to rebase my branch was out of date and the remote was ahead, hope that's ok. I've just copied one of the tests and added couple of cases for the new parameters.

File: modules/virtualnetwork/variables.tf (hunk 5 above)
Comment (b0bu) [reply modules/virtualnetwork/variables.tf:68]: Just getting to this, it's a nested object within the `virtual_networks` object. Can you do
```hcl
    vwan_security_configuration = optional(object({
      secure_internet_traffic = optional(bool, false)
      secure_private_traffic  = optional(bool, false)
    }), {})
```

Comment (b0bu) [reply modules/virtualnetwork/variables.tf]: turns out you can :D

**Comment by matt-FFFFFF** [comment]
Thanks @b0bu the tests look good - will get this merged!

**Review by matt-FFFFFF — APPROVED** [review]

**Comment by b0bu** [comment]
> Thanks @b0bu the tests look good - will get this merged!

aweseome, sorry for the radio silence, last minute paternity leave!
